### DISCUSSION TITLE
[Question] uamethod no parent

### DISCUSSION THREAD

**Discussion by AiyionPrime** [FreeOpcUa/opcua-asyncio#1496] [Q&A] [answered]
Good afternoon everyone;
I came across uamethod these days and wanted to know whether it would be possible to use it as decorator for functions whose signature does not have "parent" as first attribute?

https://github.com/FreeOpcUa/opcua-asyncio/blob/e9978c157d6c68720143211b9d794f2a8cba245e/asyncua/common/methods.py#L73-L90

When the fuction decorated with it is actually a method and the class is already the parent of the "function" I see little use in having it being part of every of our server methods?

If I understand it correctly a method currently can't look like this, if I want to decorate it with uamethod, even if it' not making use of 'parent'?

```python
from asyncua.common import uamethod
from asyncua.sync import Server

class PinController:

    def __init__(self, some_server: Server):
        self.server = some_server
        server.link_method(
            server.get_node(ua.NodeId.from_string("ns=1;i=42")), uamethod(self.on_set_output)
        )

    def on_set_output(self, pin: ua.UInt32, value: ua.Boolean) -> None:
        # set the pin to something useful
        pass
```

**Comment by AndreasHeine** [comment]
ua methods are always part of an opc ua objecttype so the always have a parent!

**Comment by AndreasHeine** [comment]
you probably need to create a own Pincontrollertype, instantiate a instance of that type and bind the instances methods to class functions

**Comment by AiyionPrime** [comment]
Maybe my example was misleading. If I were to implement an OPC UA Server based on PinController, and wanted the method `on_set_output` to be the OPC UA "function" that handles the request, then it is nice, that the server gives a parent attribute in it's direction, but I don't want it.

Lets see. The Class above could have a reference to a server like this and link methods accordingly.

I've updated the example above; maybe my case became somewhat clearer. Sorry for the confusion.

**Comment by schroeder-** [comment, marked as answer]
Just make a wrapper function

```
 def remove_parent(func):
     def wrap(parent, *args):
          return func(*args)
     return wrap

class PinController:

    def __init__(self, some_server: Server):
        self.server = some_server
        server.link_method(
            server.get_node(ua.NodeId.from_string("ns=1;i=42")), uamethod(remove_parent(self.on_set_output))
        )

    def on_set_output(self, pin: ua.UInt32, value: ua.Boolean) -> None:
        # set the pin to something useful
        pass
```

**Reply by AiyionPrime** [reply]
I did so; just wondered whether this was unidiomatic. Thanks :)
